vortex-array/src/validity.rs

```diff
@@ -211,6 +212,76 @@ impl Validity {
         Ok(validity)
     }
 
+    pub fn patch<P: AsPrimitive<usize>>(
+        self,
+        len: usize,
+        positions: &[P],
+        patches: Validity,
+    ) -> VortexResult<Self> {
+        if positions.last().map(|p| p.as_() >= len).unwrap_or(false) {
+            vortex_bail!("Position must be within length")
+        }
+
+        match (self, patches) {
```

Comment:
This block has pretty high cognitive complexity because you have to handle all 16 permutations.

I strongly suspect it'd be simpler if you did this, which just has two match statements with 4 branches each:

```
if matches!(self, Validity::NonNullable | Validity::AllValid) && matches!(patches, Validity::NonNullable | Validity::AllValid) || self.eq(patches) {
    return Ok(self);
}

let base = match self {
    ...
}
let patches = match patches {
   ...
}
Validity::try_from(base.patch(positions, patches)?.into_array())
```

Reply:
yeah, I focused on getting it right and didn't simplify at the end